src/utils/simulation.js

```diff
@@ -0,0 +1,210 @@
+   * @param {number} count - The amount of people we want from the alive array.
+   * @returns {array} sample - Person objects who can get sick.
+   */
+  interaction_sample (alive, count) {
+    let sample = [];
+    let counter = 0; 
+
+    while (counter < count) {
+      const random_num = Math.floor(Math.random() * alive.length);
+      const alive_person= alive[random_num];
+      // checks sample array to see if alive person is not in array
+      if (!sample.some(person =>  person === alive_person)) {
+        sample.push(alive_person);
+        counter += 1
+      }
+    }
+    return sample;
+  }
+
+  /**
+   * Each infected person is passed to the method alivePeople to generate a list of people that can still get infected.
+   * If healthy people is less than the total interactions, the infected person interacts will all healthy people. 
+   * If healthy people is more than the total interactions, we generate an array with the amount needed then infected
+   * person interacts with each one. 
+   */
+  time_step() {
+    // console.log("working on time step & interactions")
+    this.current_infected.forEach((infected) => {
+      // list of people that can get infected
+      let healthy_people = this.alivePeople(infected);
+      // if less than total interactions needed => interact with all
+      if (healthy_people.length <= this.total_interactions) {
+        for (let i = 0; i < healthy_people.length; i++) {
+          const healthy_person = healthy_people[i];
+          // check that person was not previously interact and infected during same timestep
+          if (this.newly_infected.indexOf(healthy_person._id) === -1) {
+            this.interaction(healthy_person._id);
+          } 
+        }
+      } else {
+        // get interaction sample with total_interaction amount 
+        const interactions = this.interaction_sample(healthy_people, this.total_interactions);
+        interactions.forEach((rand_healthy_person) => {
+          // if random person is not already in newly_infected => interaction
+          if (this.newly_infected.indexOf(rand_healthy_person._id) === -1) {
+            this.interaction(rand_healthy_person._id);
+          }  
+        })
+      }
+    })
+  }
+
+  /**
+   * @param id - The healthy unvaccinated person's id that interacted with the infected person.
+   * Called any time an infected person interacts with a healthy unvaccinated person. 
+   * If random person becomes infected, their id is then added to newly_infected array. 
+   */
+  interaction(id) {
+    const random_num = Math.random();
+    if (random_num < this.virus.repro_rate) {
+      this.newly_infected.push(id);
+    }
+  }
+
+  /**
+   * Iterates through the list of ids stored in newly_infected and updates
+   * each Person object with the virus object.
+   * Newly_infected is set to current_infected and then to an empty array. 
+   */
+  infect_newly_infected() {
+    for (let i = 0; i < this.newly_infected.length; i++) {
```

Comment:
nit: this could be made simpler with forEach

Reply:
Changed into forEach()